tensor_comprehensions/__init__.py

```diff
@@ -12,20 +12,372 @@
+                                 *inputs: torch.Tensor):
+            already_compiled = tc_obj.compilation_cache.is_compiled(
+                entry_point, inputs)
+
+            if already_compiled:
+                return
+
+            if not SILENT:
+                print(
+                    "TC \"{}\" was not explicitly compiled for ".format(entry_point) +
+                    "inputs of sizes:\n  {}\n".format("".join("{}, ".format(
+                        i.size().__str__()) for i in inputs)) +
+                    "....Generate implicit MappingOptions")
+
+            mapping_options = tc_obj.mapping_options_factory(
+                tc_obj.tc, entry_point, *inputs)
+
+            assert mapping_options is not None, (
+                "No options found for TC {} ".format(entry_point) +
+                "with inputs of sizes:\n  {}\n".format(
+                    "".join("{} ".format(i.size().__str__()) for i in inputs)))
+
+            # Compile best options to set the executor for the current
+            #     (entry point, inputs)
+            start = time.clock()
+            tc_obj.compilation_cache.compile(
+                entry_point, inputs, mapping_options)
+            if not SILENT:
+                print(
+                    "Done compiling TC \"{}\" (compile time: {}ms)".format(
+                        entry_point, int((time.clock() - start) * 10 ** 3)))
+
+        __implicit_compile__(self, entry_point, *inputs)
+
+        inputs = tuple(inputs) if len(inputs) > 1 else tuple(inputs, )
+        if unchecked:
+            return self.compilation_cache.unchecked_run(
+                entry_point, inputs, outputs)
+
+        return self.compilation_cache.run(entry_point, inputs, outputs)
+
+    def set_mapping_options_factory(self, mapping_options_factory):
+        self.mapping_options_factory = mapping_options_factory
+
+def define(tc, mapping_options_factory):
+    return TC(tc, mapping_options_factory)
+
+class TcFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, forward_fun, backward_fun, *inputs):
+        ctx.backward_fun = backward_fun
+        ctx.save_for_backward(*inputs)
+        return tuple(forward_fun(*inputs))
+
+    @staticmethod
+    def backward(ctx, *gradients):
+        if ctx.backward_fun is not None:
+            inputs = tuple(ctx.saved_tensors) + tuple(
+                t.contiguous() for t in gradients)
+            # PyTorch convention: need an extra None return for each of
+            # forward_fun and backward_fun,
+            return (None, None, *tuple(ctx.backward_fun(*inputs)))
+
+        return (None, )
+
+class TCWithAutograd(object):
+    def __init__(self, forward_fun, backward_fun):
+        self.forward_fun = forward_fun
+        self.backward_fun = backward_fun
+
+    def __call__(self, *inputs):
+        return TcFunction.apply(self.forward_fun, self.backward_fun, *inputs)
+
+def define_with_autograd(forward_fun, backward_fun):
```

Comment:
the naming implies there is some relation with `define`, but there is very little here